### PR TITLE
Eb2

### PR DESCRIPTION
**Pull request by enkhbateddie**
Have a look, guys. Have you got any particular questions write comments at any section pls?

### CODE REVIEW

File: src/bcccp/carpark/paystation/PaystationController.java [hunk 1]
```diff
@@ -6,76 +6,143 @@
...
 
 	@Override
 	public void ticketTaken() {
-		// TODO Auto-generated method stub
-		
-		if(state == State.WAITING || state == State.PAID || state == State.REJECTED ) {
-			state = State.IDLE;
-		}else {
-			beep();
+		if (state_ == STATE.IDLE) {
```
Comment (enkhbateddie) [inline src/bcccp/carpark/paystation/PaystationController.java:135]: Where is REJECTED state?

File: src/bcccp/carpark/exit/ExitController.java [hunk 2]
```diff
@@ -6,149 +6,320 @@
 import bcccp.carpark.ICarpark;
 import bcccp.carpark.IGate;
 import bcccp.tickets.adhoc.IAdhocTicket;
+import bcccp.carpark.ICarSensor;
```
Comment (enkhbateddie) [inline src/bcccp/carpark/exit/ExitController.java:9]: import bcccp.carpark.ICarSensor;
+import bcccp.carpark.ICarSensorResponder;
 +import bcccp.carpark.ICarpark;
 +import bcccp.carpark.IGate;
 +import bcccp.tickets.adhoc.IAdhocTicket;
All good imports.

File: src/bcccp/carpark/exit/ExitController.java [hunk 3]
```diff
@@ -6,149 +6,320 @@
 import bcccp.carpark.ICarpark;
 import bcccp.carpark.IGate;
 import bcccp.tickets.adhoc.IAdhocTicket;
```
Comment (enkhbateddie) [inline src/bcccp/carpark/exit/ExitController.java:8]: Need some imports:
import bcccp.carpark.ICarSensor;
+import bcccp.carpark.ICarSensorResponder;
+import bcccp.carpark.ICarpark;
+import bcccp.carpark.IGate;
+import bcccp.tickets.adhoc.IAdhocTicket;

File: src/bcccp/carpark/paystation/PaystationController.java [hunk 4]
```diff
@@ -6,15 +6,16 @@
...
-	enum State
-    {
-		IDLE, WAITING, REJECTED, PAID
-    }
-	private IPaystationUI ui;	
-	private ICarpark carpark;
-	protected State state;
-	private IAdhocTicket  adhocTicket = null;
-	private float charge;
+private enum STATE { IDLE, WAITING, REJECTED, PAID }
```
Comment (enkhbateddie) [inline src/bcccp/carpark/paystation/PaystationController.java:9]: private enum STATE declaration good choice.

File: src/bcccp/carpark/paystation/PaystationController.java [hunk 5]
```diff
@@ -6,15 +6,16 @@
 public class PaystationController 
 		implements IPaystationController {
 	
-	enum State
```
Comment (enkhbateddie) [inline src/bcccp/carpark/paystation/PaystationController.java:9]: is it protected or private?

File: src/bcccp/carpark/paystation/PaystationController.java [hunk 6]
```diff
@@ -20,14 +20,12 @@
...
-		this.ui = ui;
-		this.state = State.IDLE;
-		}catch(NullPointerException e) {
-			display("Can't be null");
-		}
+		
+		this.carpark_ = carpark;
+		this.ui_ = ui;
+		
+		ui.registerController(this);
```
Comment (enkhbateddie) [inline src/bcccp/carpark/paystation/PaystationController.java:30]: Where is the exception?

File: src/bcccp/carpark/paystation/PaystationController.java [hunk 7]
```diff
@@ -28,6 +28,43 @@ public PaystationController(ICarpark carpark, IPaystationUI ui) {
 		setState(STATE.IDLE);		
 	}
 
+	private void log(String message) {
+		System.out.println("EntryController : " + message);
+	}
+
+	
+	
+	private void setState(STATE newState) {
```
Comment (enkhbateddie) [inline src/bcccp/carpark/paystation/PaystationController.java]: Totally new! good method.

File: src/bcccp/carpark/paystation/PaystationController.java [hunk 8]
```diff
@@ -93,16 +93,25 @@ public void ticketInserted(String barcode) {
 
 	@Override
 	public void ticketPaid() {
-		// TODO Auto-generated method stub
+		if (state_ == STATE.WAITING) {
```
Comment (enkhbateddie) [inline src/bcccp/carpark/paystation/PaystationController.java:107]: Quite long

File: src/bcccp/carpark/exit/ExitController.java (hunk 3 above)
Comment (haileydo) [reply src/bcccp/carpark/exit/ExitController.java:8]: maybe you can simplify as 
```
import bcccp.carpark.*;
import bcccp.tickets.*;
```

File: src/bcccp/carpark/exit/ExitController.java [hunk 9]
```diff
@@ -6,149 +6,320 @@
...
-		IDLE, WAITING, BLOCKED, PROCESSED, REJECTED, TAKEN, EXITING, EXITED
-    }
+implements ICarSensorResponder,
+           IExitController {
+
+private enum STATE { IDLE, WAITING, PROCESSED, REJECTED, TAKEN, EXITING, EXITED, BLOCKED } 
+
+private STATE state;
+private STATE prevState;
+private String message;
```
Comment (haileydo) [inline src/bcccp/carpark/exit/ExitController.java:24]: maybe change the private variable names according to convention of underscore?

```
private STATE state_;
private STATE prevState_;
private String message_;
```

File: src/bcccp/carpark/paystation/PaystationController.java (hunk 6 above)
Comment (haileydo) [reply src/bcccp/carpark/paystation/PaystationController.java:30]: probably you can throw an exception here before the setting up of constructors:
```
if (carpark == null || ui == null) {
throw new IllegalArgumentException("Parameters can't be null);
}
```

File: src/bcccp/carpark/paystation/PaystationController.java [hunk 10]
```diff
@@ -6,76 +6,143 @@
...
+				ui_.display("Take Rejected Ticket");
+				setState(STATE.REJECTED);
+				log("ticketInserted: ticket is not current");				
+			}
+		}
+		else {
+			ui_.beep();
+			log("ticketInserted: called while in incorrect state");				
+		}
 	}
```
Comment (haileydo) [inline src/bcccp/carpark/paystation/PaystationController.java:98]: May consider using switch?
```
switch(state_):
case IDLE:
adhocTicket_ = carpark_.getAdhocTicket(barcode);
if (adhocTicket_ = null) {
throw new NullPointerException("Can't find barcode");
}
else {
charge_ = carpark_.calculateAddHocTicketCharge(adhocTicket_.getEntryDateTime());
ui_.display("Pay " + String.format("%.2f", charge_));
setState(STATE.WAITING);
break;
}
default:
ui_.beep();
setState(STATE.REJECTED);
log("ticketInserted: called while in incorrect state");

```

File: src/bcccp/carpark/paystation/PaystationController.java (hunk 8 above)
Comment (haileydo) [reply src/bcccp/carpark/paystation/PaystationController.java:107]: should be ok

File: src/bcccp/carpark/paystation/PaystationController.java (hunk 1 above)
Comment (haileydo) [reply src/bcccp/carpark/paystation/PaystationController.java:135]: similar, you can consider using switch and default case

File: src/bcccp/carpark/paystation/PaystationController.java (hunk 4 above)
Comment (ankushgdb) [reply src/bcccp/carpark/paystation/PaystationController.java:9]: yes it is. did you try to execute this code?

File: src/bcccp/carpark/paystation/PaystationController.java [hunk 11]
```diff
@@ -6,15 +6,16 @@
...
+private enum STATE { IDLE, WAITING, REJECTED, PAID } 
+	
+	private STATE state_;
+	
+	private IPaystationUI ui_;
+	
+	private ICarpark carpark_;
+
+	private IAdhocTicket  adhocTicket_ = null;
+	private float charge_;
```
Comment (ankushgdb) [inline src/bcccp/carpark/paystation/PaystationController.java]: correct way to define private variables

File: src/bcccp/carpark/exit/ExitController.java (hunk 9 above)
Comment (ankushgdb) [reply src/bcccp/carpark/exit/ExitController.java:24]: yes try that one please to resolve the error of private variable

File: src/bcccp/carpark/paystation/PaystationController.java (hunk 5 above)
Comment (haileydo) [reply src/bcccp/carpark/paystation/PaystationController.java:9]: should be public

**Pull request merged**